### PR TITLE
Introduce index option named 'index.percolator.map_unmapped_fields_as_string'

### PR DESCRIPTION
**Pull request by sweetest**
Related issues : #9025, #9053 

Please review this and merge this to the master.

### CODE REVIEW

**Comment by martijnvg** [comment]
Hi @sweetest thanks for the PR! I looked at it and I see that you basically brought back the `index.percolator.allow_unmapped_fields` option, but then gave it a different name. 

I think this setting should be implemented differently to actually map unknown fields as the strings. I think that something like this should be done:
1) Add a new boolean field called `mapUnmappedFieldsAsString` to QueryParseContext.java, by default this setting should be false, because this behaviour may only be enabled for the percolator. 
2) Change the failIfFieldMappingNotFound method to be something like this:
```java
private MapperService.SmartNameFieldMappers failIfFieldMappingNotFound(String name, MapperService.SmartNameFieldMappers fieldMapping) {
        if (allowUnmappedFields) {
            return fieldMapping;
        } else if (mapUnmappedFieldsAsString) {
            StringFieldMapper.Builder builder = MapperBuilders.stringField(name);
            StringFieldMapper stringFieldMapper = builder.build(new Mapper.BuilderContext(ImmutableSettings.EMPTY, new ContentPath(1)));
            return new MapperService.SmartNameFieldMappers(mapperService(), new FieldMappers(stringFieldMapper), null, false);
        } else {
            Version indexCreatedVersion = indexQueryParser.getIndexCreatedVersion();
            if (fieldMapping == null && indexCreatedVersion.onOrAfter(Version.V_1_4_0_Beta1)) {
                throw new QueryParsingException(index, "Strict field resolution and no field mapping can be found for the field with name [" + name + "]");
            } else {
                return fieldMapping;
            }
        }
    }
```
3) Remove QueryParseContext#smartNameFieldMapper(String) method. (it is not used)
4) In PercolatorQueriesRegistry.java make sure that the QueryParseContext is constructed with `mapUnmappedFieldsAsString` set to true if the `index.percolator.map_unmapped_fields_as_string` setting is enabled. 

Does this make sense?

**Comment by sweetest** [comment]
@martijnvg Thanks for the review, I modified and committed again considering your comment. Please review if there's anything more to be changed.

File: src/test/java/org/elasticsearch/percolator/PercolatorBackwardsCompatibilityTests.java [hunk 1]
```diff
@@ -60,15 +63,30 @@ public void testPercolatorUpgrading() throws Exception {
...
         } catch (PercolatorException e) {
             e.printStackTrace();
             assertThat(e.getRootCause(), instanceOf(QueryParsingException.class));
         }
-    }
 
+        // If and only if index.percolator.map_unmapped_fields_as_string is set to true, unmapped field in query is allowed.
+        ImmutableSettings.Builder settings = ImmutableSettings.settingsBuilder()
+                .put(indexSettings())
+                .put("index.percolator.map_unmapped_fields_as_string", true);
```
Comment (martijnvg) [inline src/test/java/org/elasticsearch/percolator/PercolatorBackwardsCompatibilityTests.java]: The `index.percolator.map_unmapped_fields_as_string` is not a bwc thing. I think it is better to move the testing of this setting to the PercolatorTests class.

File: src/main/java/org/elasticsearch/index/query/QueryParseContext.java [hunk 2]
```diff
@@ -392,9 +390,17 @@ public void setAllowUnmappedFields(boolean allowUnmappedFields) {
...
+    public void setMapUnmappedFieldAsString(boolean mapUnmappedFieldAsString) {
+        this.mapUnmappedFieldAsString = mapUnmappedFieldAsString;
+    }
+
+    private MapperService.SmartNameFieldMappers failIfFieldMappingNotFound(String name, MapperService.SmartNameFieldMappers fieldMapping) {
         if (allowUnmappedFields) {
             return fieldMapping;
+        } else if (mapUnmappedFieldAsString){
+            StringFieldMapper.Builder builder = MapperBuilders.stringField(name);
+            StringFieldMapper stringFieldMapper = builder.build(new Mapper.BuilderContext(ImmutableSettings.EMPTY,new ContentPath(1)));
```
Comment (martijnvg) [inline src/main/java/org/elasticsearch/index/query/QueryParseContext.java]: Maybe add a space between the two arguments to BuilderContext constructor?

File: src/main/java/org/elasticsearch/index/percolator/PercolatorQueriesRegistry.java [hunk 3]
```diff
@@ -208,7 +213,11 @@ private Query parseQuery(String type, XContentParser parser) {
             // Query parsing can't introduce new fields in mappings (which happens when registering a percolator query),
             // because field type can't be inferred from queries (like document do) so the best option here is to disallow
             // the usage of unmapped fields in percolator queries to avoid unexpected behaviour
+            //
+            // For backward compatibility, query can contain unmapped fields only if index.percolator.map_unmapped_fields_as_string
+            // is set to true
```
Comment (martijnvg) [inline src/main/java/org/elasticsearch/index/percolator/PercolatorQueriesRegistry.java]: I think this bwc comment can be removed? The `index.percolator.map_unmapped_fields_as_string` setting is there just to facilitate in the case a field is unmapped and then automatically creating a default string field.

File: src/test/java/org/elasticsearch/percolator/PercolatorBackwardsCompatibilityTests.java (hunk 1 above)
Comment (martijnvg) [inline src/test/java/org/elasticsearch/percolator/PercolatorBackwardsCompatibilityTests.java]: Just add a new test case, that specifically expects no failure when a field mentioned in a percolator does not exist.

**Comment by martijnvg** [comment]
@sweetest The change looks good! I left some more comments, mainly regarding the test.

**Comment by sweetest** [comment]
@martijnvg I've committed again according to your comments, please check when possible.

File: src/test/java/org/elasticsearch/percolator/PercolatorTests.java [hunk 4]
```diff
@@ -2034,5 +2029,25 @@ public void testNestedDocFilter() throws IOException {
...
+    @Test
+    public void testMapUnmappedFieldAsString() throws IOException{
+        // If index.percolator.map_unmapped_fields_as_string is set to true, unmapped field is mapped as an analyzed string.
+        ImmutableSettings.Builder settings = ImmutableSettings.settingsBuilder()
+                .put(indexSettings())
+                .put("index.percolator.map_unmapped_fields_as_string", true);
+        assertAcked(prepareCreate("test")
+                .setSettings(settings));
+        client().prepareIndex("test", PercolatorService.TYPE_NAME)
+                .setSource(jsonBuilder().startObject().field("query", queryStringQuery("VALUE")).endObject()).get();
```
Comment (martijnvg) [inline src/test/java/org/elasticsearch/percolator/PercolatorTests.java]: maybe just use a match query on field `field1` with value `value`, to make it clearer that this is query on field, that would normally fail?

File: src/test/java/org/elasticsearch/percolator/PercolatorBackwardsCompatibilityTests.java [hunk 5]
```diff
@@ -22,15 +22,18 @@
 import org.elasticsearch.action.index.IndexRequestBuilder;
 import org.elasticsearch.action.percolate.PercolateResponse;
 import org.elasticsearch.action.percolate.PercolateSourceBuilder;
+import org.elasticsearch.common.settings.ImmutableSettings;
```
Comment (martijnvg) [inline src/test/java/org/elasticsearch/percolator/PercolatorBackwardsCompatibilityTests.java]: Not that the test has been moved to a different class, all the modifications to this file can be undone, so it doesn't appear in the list of changed files.

**Comment by martijnvg** [comment]
@sweetest I left two small comments, other than that this change looks good to me.

**Comment by sweetest** [comment]
@martijnvg 

Applied all the comments and checked import statements again, thanks for the detailed review.

hope this to be included in after releases

**Comment by martijnvg** [comment]
Thanks @sweetest!

**Pull request closed without merging**